### PR TITLE
feat(enhancement): Acceleration and turning multiplier attributes

### PR DESCRIPTION
**Pull request by Quantumshark**
**Feature:** This PR implements the feature request detailed and discussed in issue #9234 

## Feature Details
Adds two attributes, `"acceleration multiplier"` and `"turn multiplier"`. A ship's acceleration is multiplied by `"acceleration multiplier"`, while its rate of turn is multiplied by `"turn multiplier"`.

## UI Screenshots
N/A

## Usage Examples
While this can be added to outfits, I imagine it will primarily be used on ship hulls.

## Testing Done
Bought three lightning bugs, gave one `acceleration multiplier -.9`, one `turn multiplier 4` and one `accleration multiplier 9`, expecting -90% acceleration, +400% turning and +900% acceleration respectively. Confirmed that the stats displayed as expected in every view, and that they performed as expected in-game.
Also tested a stock bulk freighter with `inertia reduction 4`  and `acceleration multiplier 4`, confirming that `inertia reduction` and `acceleration multiplier` worked correctly together and that reverse thrust worked properly.

### Automated Tests Added
none

## Performance Impact
One more attribute check when calculating turning and acceleration, each of which are already checking (directly or indirectly) at least three attributes (thrust/turn, mass, and inertia reduction). So, should be very minor unless I've done something wrong.

### CODE REVIEW

**Comment by tibetiroka** [comment]
Shouldn't we call it `thrust multiplier`? That would be more in line with existing names (and #9114, if that happens).

**Comment by Quantumshark** [comment]
A `thrust multiplier` would increase both acceleration and maximum speed. This only changes your acceleration, hence `acceleration multiplier.`

File: source/AI.cpp [hunk 1]
```diff
@@ -2177,7 +2177,7 @@ bool AI::Stop(Ship &ship, Command &command, double maxSpeed, const Point directi
 		forwardTime += stopTime;
 
 		// Figure out your reverse thruster stopping time:
-		double reverseAcceleration = ship.Attributes().Get("reverse thrust") / ship.InertialMass();
+		double reverseAcceleration = ship.ReverseAcceleration();
 		double reverseTime = (180. - degreesToTurn) / ship.TurnRate();
 		reverseTime += speed / reverseAcceleration;
```
Comment (TomGoodIdea) [inline source/AI.cpp:2182]:
```suggestion
		double reverseTime = (180. - degreesToTurn) / ship.TurnRate();
		reverseTime += speed / ship.ReverseAcceleration();
```
as `reverseAcceleration` is used only once here

File: source/Ship.cpp [hunk 2]
```diff
@@ -2700,15 +2700,17 @@ double Ship::InertialMass() const
 
 double Ship::TurnRate() const
 {
-	return attributes.Get("turn") / InertialMass();
+	return (attributes.Get("turn") / InertialMass()
+		* (1 + attributes.Get("turn multiplier")));
```
Comment (TomGoodIdea) [inline source/Ship.cpp:2704]:
```suggestion
	return attributes.Get("turn") / InertialMass()
		* (1 + attributes.Get("turn multiplier"));
```

File: source/Ship.cpp [hunk 3]
```diff
@@ -2700,15 +2700,17 @@ double Ship::InertialMass() const
...
 }
 
 
 
 double Ship::Acceleration() const
 {
 	double thrust = attributes.Get("thrust");
-	return (thrust ? thrust : attributes.Get("afterburner thrust")) / InertialMass();
+	return ((thrust ? thrust : attributes.Get("afterburner thrust")) / InertialMass()
+		* (1 + attributes.Get("acceleration multiplier")));
```
Comment (TomGoodIdea) [inline source/Ship.cpp:2713]:
```suggestion
	return (thrust ? thrust : attributes.Get("afterburner thrust")) / InertialMass()
		* (1 + attributes.Get("acceleration multiplier"));
```

File: source/Ship.cpp [hunk 4]
```diff
@@ -2726,7 +2728,8 @@ double Ship::MaxVelocity() const
 
 double Ship::ReverseAcceleration() const
 {
-	return attributes.Get("reverse thrust");
+	return ((attributes.Get("reverse thrust") / InertialMass())
+		* (1 + attributes.Get("acceleration multiplier")));
```
Comment (TomGoodIdea) [inline source/Ship.cpp:2732]:
```suggestion
	return attributes.Get("reverse thrust") / InertialMass()
		* (1 + attributes.Get("acceleration multiplier"));
```

File: source/Ship.cpp [hunk 5]
```diff
@@ -4293,7 +4296,7 @@ void Ship::DoMovement(bool &isUsingAfterburner)
 					slowness += scale * attributes.Get(isThrusting ? "thrusting slowing" : "reverse thrusting slowing");
 					disruption += scale * attributes.Get(isThrusting ? "thrusting disruption" : "reverse thrusting disruption");
 
-					acceleration += angle.Unit() * (thrustCommand * thrust / mass);
+					acceleration += angle.Unit() * (thrustCommand * (isThrusting ? Acceleration() : ReverseAcceleration()));
```
Comment (TomGoodIdea) [inline source/Ship.cpp:4299]:
```suggestion
					acceleration += angle.Unit() * thrustCommand * (isThrusting ? Acceleration() : ReverseAcceleration());
```
The parentheses are in master too, but they don't seem really needed.

File: source/Ship.cpp [hunk 6]
```diff
@@ -4347,7 +4350,8 @@ void Ship::DoMovement(bool &isUsingAfterburner)
 	if(acceleration)
 	{
 		acceleration *= slowMultiplier;
-		Point dragAcceleration = acceleration - velocity * (Drag() / mass);
+		// Acceleration multiplier needs to modify effective drag, otherwise it changes top speeds.
+		Point dragAcceleration = acceleration - velocity * (Drag() * (1 + attributes.Get("acceleration multiplier")) / mass);
```
Comment (TomGoodIdea) [inline source/Ship.cpp:4354]:
```suggestion
		Point dragAcceleration = acceleration - velocity * Drag() * (1 + attributes.Get("acceleration multiplier")) / mass;
```
same here

File: source/ShipInfoDisplay.cpp [hunk 7]
```diff
@@ -258,18 +258,24 @@ void ShipInfoDisplay::UpdateAttributes(const Ship &ship, const PlayerInfo &playe
 	fullMass /= reduction;
 	attributeLabels.push_back("acceleration:");
 	if(!isGeneric)
-		attributeValues.push_back(Format::Number(3600. * forwardThrust / currentMass));
+		attributeValues.push_back(Format::Number((3600. * forwardThrust / currentMass)
```
Comment (TomGoodIdea) [inline source/ShipInfoDisplay.cpp:261]:
```suggestion
		attributeValues.push_back(Format::Number(3600. * forwardThrust / currentMass
```

File: source/ShipInfoDisplay.cpp [hunk 8]
```diff
@@ -258,18 +258,24 @@ void ShipInfoDisplay::UpdateAttributes(const Ship &ship, const PlayerInfo &playe
...
+			* (1 + attributes.Get("acceleration multiplier"))));
 	else
-		attributeValues.push_back(Format::Number(3600. * forwardThrust / fullMass)
-			+ " - " + Format::Number(3600. * forwardThrust / emptyMass));
+		attributeValues.push_back(Format::Number((3600. * forwardThrust / fullMass)
+			* (1 + attributes.Get("acceleration multiplier")))
+			+ " - " + Format::Number((3600. * forwardThrust / emptyMass)
```
Comment (TomGoodIdea) [inline source/ShipInfoDisplay.cpp:266]:
```suggestion
		attributeValues.push_back(Format::Number(3600. * forwardThrust / fullMass
			* (1 + attributes.Get("acceleration multiplier")))
			+ " - " + Format::Number(3600. * forwardThrust / emptyMass
```

File: source/ShipInfoDisplay.cpp [hunk 9]
```diff
@@ -258,18 +258,24 @@ void ShipInfoDisplay::UpdateAttributes(const Ship &ship, const PlayerInfo &playe
...
+			* (1 + attributes.Get("acceleration multiplier"))));
 	attributesHeight += 20;
 
 	attributeLabels.push_back("turning:");
 	if(!isGeneric)
-		attributeValues.push_back(Format::Number(60. * attributes.Get("turn") / currentMass));
+		attributeValues.push_back(Format::Number((60. * attributes.Get("turn") / currentMass)
+			* (1 + attributes.Get("turn multiplier"))));
```
Comment (TomGoodIdea) [inline source/ShipInfoDisplay.cpp:273]:
```suggestion
		attributeValues.push_back(Format::Number(60. * attributes.Get("turn") / currentMass
			* (1 + attributes.Get("turn multiplier"))));
```

**Comment by Amazinite** [comment]
Except mass affects both acceleration and turn and thrust affects both acceleration and max speed, which would leave us at square 1 of the problem described in #9234.

**Comment by Amazinite** [comment]
Why? The confounding variable between acceleration and max speed isn't drag, it's thrust. Drag is only used for max speed, so changing the max speed is very easy. In what way does thrust affecting acceleration and max speed and mass affecting acceleration and turn rate mean that drag is what should change?

**Comment by Quantumshark** [comment]
There's no sense in which calling an attribute that multiplies turn rate `moment of inertia` is more intuitive than calling it `turn multiplier`.

Being able to independently tweak a ship's turning and acceleration would be useful, and would provide more room to differentiate ship handling. The issue we have currently is that acceleration is the result of two factors, thrust and mass, each of which also affects something else (thrust affects top speed, mass affects turn rate). Yes, it would be possible to achieve any desired set of outputs (top speed, acceleration, and turn rate) by configuring the four inputs (thrust, mass, drag and turn). Doing so would, however, be very convoluted, and would likely mess things up elsewhere. Alternatively, we can use these two simple, intuitive attributes to achieve full control over ship mobility, making it far easier to balance things.

**Comment by Zitchas** [comment]
Just a thought: There is another way to solve this: Change the formula so that max speed is no longer defined by anything in any of the other formulas. 

We really have no clear definition/idea of what drag even is, or why we have max speeds at all, or what it means; other than the meta element that with our rates of acceleration, having infinite max speed would cause a lot of problems. So we have a meta reason for it to exist, but no in-game lore reason for it to exist. 

Some ideas as to what our max speed actually represents:
1. The maximum speed that our ship's nav computer can handle the work of flying the ship. (fast ships have more computers to handle this work)
2. The maximum speed that our ship can ignore the effects of micrometeriorites and miscellaneous space debris. (fast ships have better deflective capabilities)
3. The maximum stress from general flight, impacting things, other forces, etc. that our ship can withstand before risking damage. (fast ships have stronger structural elements. Afterburners are pushing our ships into the safety margin and presumably to within a hairsbreadth of destruction.)
4. Space is filled with a mysterious thing that resists motion, and thus our maximum speed is dictated by our engine's total capability of warping this thing, which happens to mirror their thrust. Afterburners allow us to warp it even more. Some ships are better able to warp it than others, and this ability to warp it is measured in "drag" where lower numbers are better able to warp this thing. For unknown reasons small ships are usually better able to warp this thing than others.

All of these have problems, particularly since most of these have a very dubious connection to the engines. Sadly, option 4 is the one that comes closest to actually representing what we see in game.

If we know what max speed is actually supposed to be, then we might be able to better decide on what should go into calculating it and how it should be modifiable.

From a simplistic side, we *could* just strip thrust out entirely, and have max speed be directly determined by drag and/or universal constants. For instance, something like `1000/drag = max speed` with drag values recalibrated to generally all be in the 0 to 10 range, with things with 10 being the slowest of the slowest, and 0.1 being blindingly fast. Then ship A would always have a max speed of X. Putting in the best engines possible just means it can reach X faster, and instead opting for the most pathetic tiny ion thrusters would just mean that it takes so much longer to reach X. It would also mean that, if we want the player to be able to change their max speed, we can implement means of doing that which don't touch anything else about manoeuvring. 

That way max speed is determined by the ship, and it's ability to accelerate and reach that top speed is determined by the engines. This decoupling of max speed and acceleration would mean that we could actually have ships that are capable of very high max speeds but still have poor acceleration; and vice versa. Better ship differentiation, less vulnerable to being broken. 

Beyond this, we could figure out some other more complex means of doing so. Another idea:
a. max speed is determined based on the ratio of engine space to mass. So if the player wants a higher max speed, they need to figure out how to reduce the mass of their ship. This would give people a direct incentive to *not* use every last iota of outfit space, since empty space would have a direct benefit.

**Comment by Zitchas** [comment]
Delta has significantly reduced the impact of "mysterious space drag" as it affects flightpath and velocity handling (depending on user setting, these effects have either been minimized or shifted over to being handled transparently by lateral thrust to the best of the ship's capability), but nothing has been changed yet in regards to the max speed formula or the spaceship drag attribute.

**Comment by Quantumshark** [comment]
To clarify, a significant portion of my goal here is being able to vary acceleration more between different ship sizes - so that larger ships take longer to reach their top speed, and feel heavier, while smaller ships can make quick major changes in their speed and heading, helping them dodge. Redesigning the way top speed is calculated is an interesting concept, but it doesn't functionally serve as an alternative to this.

**Comment by Quantumshark** [comment]
> I'm not sure why this would be the case? I don't know what the desire is for the turn rates of large ships, but changing it could be done by anything from adjusting the scaling of turning thruster values to altering their masses or moment of inertias. Nothing about that means they would need to have lower moments of inertia.

The desire is for turn rates to vary significantly less with size than acceleration rates do.
Adjusting the scaling of turning values was the first approach I thought of, but it creates problems.
For one, it makes it harder to get an 'ideal' turn rate on your ship when outfitting, because the gap between different sizes is much larger, while the amount added by a smaller steering outfit becomes much smaller. It also indirectly guts the turn rate of the larger Coalition ships, since they're now having to make do with 
Altering masses alone isn't an option, because that has exactly the same effect on turning as it does on acceleration - the entire point of this PR is to fix that.

Increasing masses and using turn multipliers rather than leaving masses unchanged and using acceleration multipliers is definitely an option - which is best depends whether we want the other implications of increased mass - which are reduced vulnerability to hit force (seems sensible to couple with the lower acceleration), increased maximum heat (indirect heat damage nerf; plus a very slight effective buff to the outfit space of larger ships as they don't need as much cooling), and reduced effect of carried ships on handling (this is extremely desirable, and in my opinion the most significant of these points). So with that in mind, changing masses and adding `turn multiplier` is probably a better course of action than just adding `acceleration multiplier`, unless there's a reason otherwise that I've missed.
Whichever we use in practice, I don't think there's any problem with providing the other for use by plugin creators and potential future content.

Now, you could decide that instead of `turn multiplier`, you're instead going to have the attribute that modifies turn rates be `moment of intertia`. Presumably, a higher `moment of inertia` makes a ship turn more slowly. Thus, to replicate the effects described above, you'd have to have lower `moment of inertia` on larger ships than on smaller ones.

**Comment by Quantumshark** [comment]
> I'm not sure why this is the case? There's no reason Coalition engines need to inherently be worse in terms of turn/ton.

If we make steering scale more steeply, then the thrust/ton difference between smaller and larger steering outfits becomes larger - that's what is meant by "make steering scale more steeply". So, that would result in smaller engines having lower turn/ton compared to larger ones. This hurts the handling of Coalition ships, because their large ships are using small engines rather than large ones, and if the difference between the two is larger then large Coalition ships are going to handle noticeably worse relative to other large ships.

> Decoupling acceleration from turn rate doesn't require separate multipliers for each. Use mass to modify acceleration, and moment of inertia to modify turn rate if you need to.

Yes, it's not necessary to have both multipliers, it's just nice to support both. You could just have one or the other, and use that and mass to change things. Moment of inertia isn't meaningfully distinct from turn multiplier in this regard, either could fulfill the same role.

> Hit force makes sense to depend entirely on ship mass. We also have force protection if additional granularity on a particular ship is desired. Increased maximum heat should be handled by dedicated heat capacity modifying attributes, not by using mass to stand in for heat and using an acceleration modifier to handle kinematics. For balancing carried ships, it would make more sense to just balance their masses from the start, instead of making every other ship have an acceleration modifier to fit with them.

You've got what I'm saying the wrong way around - I'm saying that changing mass would have these effects as well as the effects on acceleration. I'm not suggesting them as things we should change for the purpose of changing them, but rather that we _can't_ change mass without them also changing.

> I'm not sure why you'd want to make large ships turn faster than small ships? If turn rate is a function of mass and the power of a ship's turning thrusters, you'd simply adjust the scaling of those to get the desired behavior. A moment of inertia attribute could be used to change this on a particular ship (ideally only if it has a special shape that would lead to interesting behavior, such as the 512), but I don't see it as being necessary to fundamental balance.

I'm not saying large ships should turn faster than small ships, I'm saying that turning should vary less with size than acceleration does. Meaning, large ships have more turn **relative to their acceleration** than small ships. That would still be less turn than small ships, but by a smaller amount than their acceleration differs by.
If you replace mass with moment of inertia for the turn calculation, then smaller ships would have more moment of inertia relative to their mass - and thus more accleration relative to their turning - while larger ships would have less moment of inertia relative to their mass, and thus more turning relative to their acceleration.

File: source/Ship.cpp [hunk 10]
```diff
@@ -4336,7 +4339,7 @@ void Ship::DoMovement(bool &isUsingAfterburner)
 				slowness += slownessCost;
 				disruption += disruptionCost;
 
-				acceleration += angle.Unit() * thrust / mass;
+				acceleration += angle.Unit() * attributes.Get("acceleration multiplier") * thrust / mass;
```
Comment (Amazinite) [inline source/Ship.cpp:4342]:
```suggestion
				acceleration += angle.Unit() * (1. + attributes.Get("acceleration multiplier")) * thrust / mass;
```

File: source/Ship.cpp [hunk 11]
```diff
@@ -4346,7 +4349,8 @@ void Ship::DoMovement(bool &isUsingAfterburner)
 	if(acceleration)
 	{
 		acceleration *= slowMultiplier;
-		Point dragAcceleration = acceleration - velocity * (Drag() / mass);
+		// Acceleration multiplier needs to modify effective drag, otherwise it changes top speeds.
+		Point dragAcceleration = acceleration - velocity * Drag() * (1 + attributes.Get("acceleration multiplier")) / mass;
```
Comment (Amazinite) [inline source/Ship.cpp:4353]:
```suggestion
		Point dragAcceleration = acceleration - velocity * Drag() * (1. + attributes.Get("acceleration multiplier")) / mass;
```

File: source/Ship.cpp [hunk 12]
```diff
@@ -2699,15 +2699,17 @@ double Ship::InertialMass() const
 
 double Ship::TurnRate() const
 {
-	return attributes.Get("turn") / InertialMass();
+	return attributes.Get("turn") / InertialMass()
+		* (1 + attributes.Get("turn multiplier"));
```
Comment (Amazinite) [inline source/Ship.cpp:2703]:
```suggestion
		* (1. + attributes.Get("turn multiplier"));
```

File: source/Ship.cpp [hunk 13]
```diff
@@ -2699,15 +2699,17 @@ double Ship::InertialMass() const
...
 }
 
 
 
 double Ship::Acceleration() const
 {
 	double thrust = attributes.Get("thrust");
-	return (thrust ? thrust : attributes.Get("afterburner thrust")) / InertialMass();
+	return (thrust ? thrust : attributes.Get("afterburner thrust")) / InertialMass()
+		* (1 + attributes.Get("acceleration multiplier"));
```
Comment (Amazinite) [inline source/Ship.cpp:2712]:
```suggestion
		* (1. + attributes.Get("acceleration multiplier"));
```

File: source/ShipInfoDisplay.cpp [hunk 14]
```diff
@@ -258,18 +258,24 @@ void ShipInfoDisplay::UpdateAttributes(const Ship &ship, const PlayerInfo &playe
 	fullMass /= reduction;
 	attributeLabels.push_back("acceleration:");
 	if(!isGeneric)
-		attributeValues.push_back(Format::Number(3600. * forwardThrust / currentMass));
+		attributeValues.push_back(Format::Number((3600. * forwardThrust / currentMass)
+			* (1 + attributes.Get("acceleration multiplier"))));
```
Comment (Amazinite) [inline source/ShipInfoDisplay.cpp:262]:
```suggestion
			* (1. + attributes.Get("acceleration multiplier"))));
```

File: source/ShipInfoDisplay.cpp [hunk 15]
```diff
@@ -258,18 +258,24 @@ void ShipInfoDisplay::UpdateAttributes(const Ship &ship, const PlayerInfo &playe
...
 	else
-		attributeValues.push_back(Format::Number(3600. * forwardThrust / fullMass)
-			+ " - " + Format::Number(3600. * forwardThrust / emptyMass));
+		attributeValues.push_back(Format::Number((3600. * forwardThrust / fullMass)
+			* (1 + attributes.Get("acceleration multiplier")))
+			+ " - " + Format::Number((3600. * forwardThrust / emptyMass)
+			* (1 + attributes.Get("acceleration multiplier"))));
```
Comment (Amazinite) [inline source/ShipInfoDisplay.cpp:267]:
```suggestion
			* (1. + attributes.Get("acceleration multiplier"))));
```

File: source/ShipInfoDisplay.cpp [hunk 16]
```diff
@@ -258,18 +258,24 @@ void ShipInfoDisplay::UpdateAttributes(const Ship &ship, const PlayerInfo &playe
...
-		attributeValues.push_back(Format::Number(60. * attributes.Get("turn") / currentMass));
+		attributeValues.push_back(Format::Number((60. * attributes.Get("turn") / currentMass)
+			* (1 + attributes.Get("turn multiplier"))));
 	else
-		attributeValues.push_back(Format::Number(60. * attributes.Get("turn") / fullMass)
-			+ " - " + Format::Number(60. * attributes.Get("turn") / emptyMass));
+		attributeValues.push_back(Format::Number((60. * attributes.Get("turn") / fullMass)
+			* (1 + attributes.Get("turn multiplier")))
+			+ " - " + Format::Number((60. * attributes.Get("turn") / emptyMass)
+			* (1 + attributes.Get("turn multiplier"))));
```
Comment (Amazinite) [inline source/ShipInfoDisplay.cpp:278]:
```suggestion
			* (1. + attributes.Get("turn multiplier"))));
```

**Review by Amazinite — CHANGES_REQUESTED** [review]
Requires updates to Outfit.cpp to allow these to reach -1 like the other multipliers, as well as updates to OutfitInfoDisplay.cpp so that the display as percentages.

**Review by Amazinite — APPROVED** [review]

**Pull request merged**